### ISSUE TITLE
Cannot specify type of return type for class constructor

### ISSUE THREAD

**Issue by morille** [google/closure-compiler#3131]
The compiler complains if the return type of a class constructor is not strictly an instance of this constructor. (Since I upgrade from the august version to v20181008).
As javascript allows to return anything from a constructor, the compiler should allow users to specify a custom return type using the "@return" doc.

Just for example if I do:
```
class Translator
{
        /**
         * Constructor with custom return type.
         *
         * @return {Promise.<Translator>}
         */
        constructor()
        {
                return this.bootstrap();
        }

        /**
         * @return {Promise.<Translator>}
         */
        async bootstrap()
        {
                return this;
        }
}
```

I don't want to get this warning:
```
Translator.js:10: WARNING - inconsistent return type
found   : Promise<Translator>
required: Translation.Translator
        return this.bootstrap();
               ^^^^^^^^^^^^^^^^
```

**Comment by lauraharker** [comment]
The compiler changed behavior in v20181008 because it switched from transpiling classes before typechecking to natively typechecking ES6 classes. 

I believe that restricting what can be returned from a class constructor simplified a lot of the typechecking logic. There are a few other places where the type system is intentionally less dynamic than the JavaScript language, in order to let the compiler do better static analysis.

@shicks (who added support for native class typechecking) is that correct?
